stuff/cron/build_problem_rec_model.py

```diff
@@ -0,0 +1,333 @@
+#!/usr/bin/python3
+
+'''Builds the user recommendation model.
+
+This script reads user submissions and builds a recommendation model for
+which problems to attempt after a user has just sucessfully solved a problem.
+
+The model is a probability distribution of which problems a random user would
+attempt to solve next, based solely on the sequence of recently solved problems
+up to the latest one.
+'''
+
+import argparse
+import logging
+import sqlite3
+import warnings
+
+from typing import List, Tuple
+
+import MySQLdb.connections
+
+import numpy as np
+import pandas as pd
+
+import lib.db
+import lib.logs
+
+
+# Default training parameters.
+_TRAIN_FRACTION = 0.8
+_NUM_FOLLOWUPS = 3
+_FOLLOWUP_DECAY = 0.4
+
+
+# Type aliases for type checking
+ProblemList = List[int]
+
+
+def mean_average_precision(predicted: ProblemList,
+                           expected: ProblemList,
+                           k: int) -> float:
+    '''Compute MAP@k - https://goo.gl/KMghXR'''
+    if not predicted or not expected:
+        return None
+    num_problems = min(len(predicted), k)
+    return sum((predicted[:num_problems] == expected[:num_problems]) *
+               (1. / np.arange(1, num_problems + 1)))
+
+
+def get_first_run_per_problem(runs: pd.DataFrame) -> pd.DataFrame:
+    '''Get ordered lists of AC runs keyed by user.
+
+    Args:
+        runs (pd.DataFrame): Unordered DataFrame with AC runs with at least
+                             identity_id, problem_id and time columns.
+    Returns:
+        pd.DataFrame indexed by identity_id with ordered list of
+        recomendations.
+    '''
+    first_ac = runs.groupby(['identity_id', 'problem_id']).apply(
+        lambda x: x.sort_values(['time']).head(1))
+    logging.info('Found %d first AC runs', len(first_ac))
+
+    return first_ac.reset_index(drop=True).groupby(['identity_id']).apply(
+        lambda x: x.sort_values(['time']))
+
+
+class TrainingConfig:
+    '''A class to store the configuration for training a model.'''
+    def __init__(self, args: argparse.Namespace):
+        self.train_fraction = args.train_fraction
+        self.num_followups = args.num_followups
+        self.followup_decay = args.followup_decay
+
+
+class Model:
+    '''A class that represents a prediction model.
+    '''
+    def __init__(self, config: TrainingConfig):
+        self.config = config
+
+        # Set of users from the training data.
+        self.users = None
```

Comment:
puedes ponerle tipos a estas variables?

```suggestion
        self.users: Optional[pd.Series] = None
```

igual con los demás.

Reply:
Done